### PR TITLE
Fix check_certificate in case of never-expire certificates

### PR DESCRIPTION
**Pull request by nametoolong**
Handle the cases where a certificate has no activation time or expiration time.

### CODE REVIEW

**Comment by danpascu** [comment]
Can you point to the standard that specifies that these fields can be missing?

**Comment by nametoolong** [comment]
[RFC 5280](https://tools.ietf.org/html/rfc5280#section-4.1.2.5) allows a certificate to have no well-defined notAfter, which would be returning (time_t)-1 in GnuTLS. Actually in GnuTLS you can even set the notBefore to -1 although you can't export such a certificate, so this PR is more of a fail-safe.

**Comment by danpascu** [comment]
According to that section in the RFC, neither of notBefore nor notAfter are optional. If you want a never expiring certificate you need to use 99991231235959Z for notAfter. The previous RFC (2459) allowed them to be optional in private keys as some sort of extension (it mentioned this in [4.2.1.4](https://tools.ietf.org/html/rfc2459#section-4.2.1.4)) but that was removed from RFC 5280. The only reference to them possibly being optional is in [appendix A.2 in RFC 5280](https://tools.ietf.org/html/rfc5280#appendix-A.2), which describes "Implicitly Tagged Module, 1988 Syntax", where it says:

```
PrivateKeyUsagePeriod ::= SEQUENCE {
     notBefore       [0]     GeneralizedTime OPTIONAL,
     notAfter        [1]     GeneralizedTime OPTIONAL }
     -- either notBefore or notAfter MUST be present
```

However this refers to private keys having them optional (PrivateKeyUsagePeriod), which was only specified in RFC 2459 but was removed from RFC 5280, so I'm inclined to believe that this part in appendix A.2 is a leftover that should have also probably be removed from RFC 5280.

The similar section that describes "Explicitly Tagged Module, 1988 Syntax" [appendix A.1 in RFC 5280](https://tools.ietf.org/html/rfc5280#appendix-A.1), lists them as mandatory for certificates:

```
Validity ::= SEQUENCE {
     notBefore      Time,
     notAfter       Time  }
```

In addition I checked gnutls, and while you can use -1 to set a never expiring date, which will result in notAfter being set to 99991231235959Z, when you read it back you do NOT get -1, but you get a valid timestamp:

```
        Validity
            Not Before: Feb 18 09:52:09 2019 GMT
            Not After : Dec 31 23:59:59 9999 GMT
```

on a 32 bit system you get:

```
In [5]: cert.activation_time
Out[5]: 1550483529

In [6]: cert.expiration_time
Out[6]: 2145914603

In [7]: datetime.utcfromtimestamp(cert.expiration_time)
Out[7]: datetime.datetime(2037, 12, 31, 23, 23, 23)
```

and on a 64 bit system you get:

```
In [5]: cert.activation_time
Out[5]: 1550483529

In [6]: cert.expiration_time
Out[6]: 253402300799

In [7]: datetime.utcfromtimestamp(cert.expiration_time)
Out[7]: datetime.datetime(9999, 12, 31, 23, 59, 59)
```

I also checked the gnutls source code and the only case where it will return -1 when getting the activation or expiration times is if it cannot read the timestamp field in the ASN.1 structure, which means it's either missing or invalid.

In conclusion, I think we can approach this in 2 ways:

1. Consider a certificate with either notBefore or notAfter missing to be invalid and reject it.
2. Accept a missing notBefore to mean that the certificate was activated as soon as it was generated.

This means that opposed to the proposed patch, I would always reject a patch that returns -1 for expiration time. As for activation time, we can do the same or be more lenient and consider a missing activation time to mean that the certificate was activated as soon as it was created.

**Comment by nametoolong** [comment]
Yes, I've made a blatant mistake. I've found out that it is a platform-dependent behavior. On Windows its return value is a signed integer (giving me some negative number) and therefore causes an exception. I will try to investigate on that.
